autorest/sender.go

```diff
@@ -295,14 +295,22 @@ func doRetryForStatusCodesImpl(s Sender, r *http.Request, count429 bool, attempt
 			return resp, err
 		}
 		delayed := DelayWithRetryAfter(resp, r.Context().Done())
-		if !delayed && !DelayForBackoffWithCap(backoff, cap, attempt, r.Context().Done()) {
+		// enforce a 2 minute cap when not counting 429 if there isn't one.
+		// this should only happen in the absense of a retry-after header.
+		if !count429 && cap == 0 {
+			cap = 2 * time.Minute
+		}
+		if !delayed && !DelayForBackoffWithCap(backoff, cap, delayCount, r.Context().Done()) {
 			return resp, r.Context().Err()
 		}
 		// when count429 == false don't count a 429 against the number
 		// of attempts so that we continue to retry until it succeeds
 		if count429 || (resp == nil || resp.StatusCode != http.StatusTooManyRequests) {
 			attempt++
 		}
+		// delay count is tracked separately from attempts to
+		// ensure that 429 participates in exponential back-off
+		delayCount++
```

Comment:
Is the TestDelayWithRetryAfterWithSuccess the only test to step through for both scenarios?

Reply:
There is also `TestDoRetryForStatusCodes_Cancel429` which tests the infinite retries on 429.